server/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java

```diff
@@ -212,14 +215,25 @@ public void testForceStaleReplicaToBePromotedToPrimary() throws Exception {
                 rerouteBuilder.add(new AllocateEmptyPrimaryAllocationCommand(idxName, shardId, storeStatus.getNode().getId(), true));
             }
         }
-        rerouteBuilder.get();
-
-        ClusterState state = client().admin().cluster().prepareState().get().getState();
 
-        Set<String> expectedAllocationIds = useStaleReplica
+        final Set<String> expectedAllocationIds = useStaleReplica
             ? Collections.singleton(RecoverySource.ExistingStoreRecoverySource.FORCED_ALLOCATION_ID)
             : Collections.emptySet();
-        assertEquals(expectedAllocationIds, state.metaData().index(idxName).inSyncAllocationIds(0));
+
+        final CountDownLatch clusterStateChangeLatch = new CountDownLatch(1);
+        final ClusterStateListener clusterStateListener = event -> {
+            final Set<String> allocationIds = event.state().metaData().index(idxName).inSyncAllocationIds(0);
+            if (expectedAllocationIds.equals(allocationIds)) {
+                clusterStateChangeLatch.countDown();
+            }
```

Comment:
Just thinking about diagnosing possible future failures, I think it would be useful to `logger.info()` any non-matching states so we can see if something weird was going on.

Reply:
addressed in 22a6ffb